zst_pointer_async/README.md

```diff
@@ -0,0 +1,357 @@
+   incorporated into a single Tock board, Tock applications must occupy minimal
+   flash and RAM space. Ideally, the abstractions required to support
+   asynchronous operation would have no impact on binary size, RAM usage, or
+   execution time.
+1. **Incremental migration from futures:** If we develop an
+   expensive-but-convenient futures API and an efficient-but-inconvenient API
+   independently, then it becomes difficult for application developers to choose
+   what API to build their applications against. Choosing the
+   efficient-but-inconvenient API for non-constrained applications wastes
+   engineering effort. Choosing the expensive-but-convenient API for a new
+   application runs the risk of requiring an application rewrite down the road
+   when the application grows large. To make this decision more reasonable, and
+   to prevent large application rewrite projects, we want to make it possible to
+   incrementally migrate a large application from the futures-based API to the
+   new API.
+
+There were a few other considerations as well:
+
+*  I tried to design the asynchronous APIs using only stable Rust features.
+*  Testability: we want to thoroughly unit-test (and occasionally
+   integration-test) all system call interfaces. This mostly means making the
+   API design dependency injection friendly.
+*  Understandability/maintainability.
+
+## ZST Pointer Design
+
+The design is based on the Tock kernel's asynchronous constructs, and keeps two
+of its major concepts:
+
+1. **A static graph of interacting state machines:** The Tock kernel has a
+   statically-specified graph of objects that interact with each other. It is
+   uncommon to use dynamic handles (which is the norm when using futures), which
+   minimizes overhead. It is normal for objects' interfaces to represent state
+   machines, and for their clients to rely on that state machine's behavior to
+   minimize the amount of duplicate state kept in RAM.
+1. **Split-phase calls:** If object `A` wants to invoke an asynchronous
+   operation implemented by object `B`, `A` calls a method on `B` to start the
+   operation, which returns without blocking, then `B` calls `A` back when the
+   operation is finished.
+
+For example, in the Tock kernel, a connection between an object `Bar` and an
+interface it depends on `Foo` would look like:
+
+```
+// The API Bar depends on.
+trait Foo {
+    fn foo(&self);
+}
+
+// Implemented by Bar; contains a callback called when the Foo::foo operation is
+// complete.
+trait FooClient {
+    fn foo_done(&self);
+}
+
+// Implemented by objects using Bar, contains a callback called when Bar's
+// operation completes.
+trait BarClient {
+    fn bar_done(&self);
+}
+
+// The `Bar` object. Note that the Foo's type is injected statically, but a
+// reference to the Foo is injected dynamically. Bar's client is injected fully
+// dynamically.
+struct Bar<F: Foo> {
+    foo: &F,  // 1 word
```

Comment:
This structure is meant to be used in a static context. Unless eliminated through dead code elimination, this will never be optimized away.

Every single statement of the form "writing X will cost Y bytes" has an implicit "unless it is entirely optimized away". I think the statement is fine as-is.